src/System.Management.Automation/engine/Attributes.cs

```diff
@@ -943,43 +970,100 @@ protected override void ValidateElement(object element)
                         Metadata.ValidateNotNullFailure);
             }
 
-            var o = element as PSObject;
-            if (o != null)
+            if(_rangeKind.HasValue)
             {
-                element = o.BaseObject;
+                // Using dynamic to handle different type of numbers such as
+                // int, double, decimal and float. 
+                dynamic val = element;
```

Comment:
Here's what i had for ```ValidateElement```.

```c#
protected override void ValidateElement(object element)
{
    if (element == null)
    {
        throw new ValidationMetadataException(
                "ArgumentIsEmpty",
                null,
                Metadata.ValidateNotNullFailure);
    }

    var o = element as PSObject;
    if (o != null)
    {
        element = o.BaseObject;
    }
    
    if (_rangeKind.HasValue)
    {

        Type commonType = GetCommonType((0).GetType(),element.GetType());

        ...
```

Then in PowerShell

```powershell
function foo { param([ValidateRange('Positive')] $bar) }

foo -1
```

At this point, element will be a string when we enter ```GetCommonType```.

What's interesting is that this does work.

```powershell
function foo { param([ValidateRange('Positive')] $bar) }

foo (-1)
```

I checked the existing code and they solve this by first getting the common type of the min and max values. They then try to cast element to this new type before comparison operations. 

Since I don't have a common type, I was thinking of trying the cast to double.